HW2/P4/driver.py

```diff
@@ -20,11 +20,25 @@ def py_median_3x3(image, iterations=10, num_threads=1):
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
 
-    for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
-        tmpA, tmpB = tmpB, tmpA
-
+    if num_threads == 1:
+        for i in range(iterations):
+            filtering.median_3x3(tmpA, tmpB, 0, 1)
+            # swap direction of filtering
+            tmpA, tmpB = tmpB, tmpA
+    else:
+        print "number of threads:", num_threads
+        for i in range(iterations):
+            threads = []
+            for th in range(num_threads):
+                onethread = threading.Thread(target=filtering.median_3x3,
+                                             args=(tmpA, tmpB, th, num_threads))
+                threads.append(onethread)
+                onethread.start()
+            for onethread in threads:
```

Comment:
Uses join to synchronize outputs. This results in less parallelization though since threads are joined sequentially.